### PR TITLE
Install with crd

### PR DESCRIPTION
**Pull request by allenlsy**
**Description**:

The purpose of this PR is to swap osm-config configmap with MeshConfig resource.

Summary of primary changes:

* osm-controller uses MeshConfig based configurator rather than configmap based.
* The default MeshConfig resource name will be `osm-mesh-config`.
* Remove unused arguments from installation workflow.
* Default MeshConfig manifest file is provided.

Planned changes in the future:

* Remove configmap check in validation webhook
* For the convenience of code review, renaming `crd_client` to `client` will be done in later changes.

Related Issue: #3039 

**Affected area**:

- New Functionality      [X]
- Install                [X]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

`No`

### CODE REVIEW

File: cmd/cli/install.go [hunk 1]
```diff
@@ -301,6 +325,29 @@ func (i *installCmd) validateOptions() error {
 		return err
 	}
 
+	if i.meshConfigPath == "" {
+		// try converting to MeshConfig from osm-config configmap
+		if meshConfig, err := i.convertOsmConfigToMeshConfig(); err != nil {
+			return fmt.Errorf("Error creating MeshConfig: %s", err)
```
Comment (eduser25) [inline cmd/cli/install.go:331]: wait, error? should this "try" error of it it does not succeed?

File: manifests/mesh-config.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    enable_privileged_init_container: false
+    log_level: error
+    max_data_plane_connections: 0
+  traffic:
+    egress: false
+    use_https_ingress: false
+    enable_permissive_traffic_policy_mode: false
+  observability:
+    enable_debug_server: true
+    prometheus_scraping: True
```
Comment (eduser25) [inline manifests/mesh-config.yaml:16]:
```suggestion
    prometheus_scraping: true
```

File: pkg/configurator/crd_methods.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	}
+	return cm, nil
+}
+
+// IsPermissiveTrafficPolicyMode tells us whether the OSM Control Plane is in permissive mode,
+// where all existing traffic is allowed to flow as it is,
+// or it is in SMI Spec mode, in which only traffic between source/destinations
+// referenced in SMI policies is allowed.
+func (c *CRDClient) IsPermissiveTrafficPolicyMode() bool {
+	return c.getMeshConfig().PermissiveTrafficPolicyMode
```
Comment (eduser25) [inline pkg/configurator/crd_methods.go:43]: unrelated: now and since you reworked all this, it'd be good to cache state so every config call doesn't have to parse the entire config map every time. We just need to update values when they change.

File: tests/e2e/e2e_helm_install_test.go [hunk 4]
```diff
@@ -15,30 +15,32 @@ var _ = OSMDescribe("Test osm control plane installation with Helm",
...
+				// 	Skip("Test is not going through InstallOSM, hence cannot be automatically skipped with NoInstall (#1908)")
+				// }
+
+				// namespace := "helm-install-namespace"
+				// release := "helm-install-osm"
+
+				// // Install OSM with Helm
+				// Expect(Td.HelmInstallOSM(release, namespace)).To(Succeed())
+
+				// configmap, err := Td.GetConfigMap("osm-config", namespace)
```
Comment (eduser25) [inline tests/e2e/e2e_helm_install_test.go:30]: I guess you are not yet done here :)

File: cmd/cli/install.go [hunk 5]
```diff
@@ -159,11 +171,19 @@ func newInstallCmd(config *helm.Configuration, out io.Writer) *cobra.Command {
 				return errors.Errorf("Could not access Kubernetes cluster, check kubeconfig: %s", err)
 			}
 			inst.clientSet = clientset
+
+			configClient, err := configClientset.NewForConfig(kubeconfig)
```
Comment (eduser25) [inline cmd/cli/install.go:175]: oh. This messed with my mind so much. Can the package import be renamed something more osm-ish? ie. `osm-config-client` or something.

File: pkg/configurator/client.go [hunk 6]
```diff
@@ -16,53 +16,6 @@ import (
 	"github.com/openservicemesh/osm/pkg/kubernetes/events"
 )
 
-const (
```
Comment (eduser25) [inline pkg/configurator/client.go:19]: Since you are not deleting `client.go` and they are on the same folder, I'd suggest you change the definitions on a different commit (you could even move them to a separate file in same folder, I'm assuming `crd_client.go` is in the same folder)

File: tests/e2e/e2e_helm_install_test.go (hunk 4 above)
Comment (allenlsy) [reply tests/e2e/e2e_helm_install_test.go:30]: Right. configmap won't be used any more. We will do helm install without the configmap template, and configure using CRD.

One more thing I want to try to test is, simulate a scenario that configmap still exist in k8s storage, and `osm install` auto converts it to CRD

File: cmd/cli/install.go (hunk 5 above)
Comment (allenlsy) [reply cmd/cli/install.go:175]: If you find it confusing, so will other people. So yes, will change the name

File: pkg/configurator/client.go (hunk 6 above)
Comment (allenlsy) [reply pkg/configurator/client.go:19]: I think we need to keep the legacy client for a while. The only use for `client.go` is during the installation, we attempt to auto convert from configmap to CRD.

Until we completely remove `client.go`, we can either keep `client.go` and `crd_client.go` (I prefer this), or:

* `client` -> `configmap_client`
* `crd_client` -> `client`

File: cmd/cli/install.go (hunk 1 above)
Comment (allenlsy) [reply cmd/cli/install.go]: Resolved. `install.go` won't be modified

File: cmd/cli/install.go [hunk 7]
```diff
@@ -159,6 +163,13 @@ func newInstallCmd(config *helm.Configuration, out io.Writer) *cobra.Command {
...
+			if err != nil {
+				return errors.Errorf("Could not access Kubernetes cluster, check kubeconfig: %s", err)
+			}
+			inst.configClient = configClient
+
```
Comment (shashankram) [inline cmd/cli/install.go:172]: Why is this required? Where is `inst.configClient` being used?

File: cmd/cli/install_test.go [hunk 8]
```diff
@@ -25,6 +27,9 @@ import (
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/kubernetes/fake"
 
+	yaml "k8s.io/apimachinery/pkg/util/yaml"
+
+	"github.com/openservicemesh/osm/pkg/apis/config/v1alpha1"
```
Comment (shashankram) [inline cmd/cli/install_test.go:32]: nit: newline between local API pkg import and other packages.

File: cmd/cli/trafficpolicy_check.go [hunk 9]
```diff
@@ -33,17 +36,20 @@ osm policy check-pods bookbuyer-client bookstore-server
...
-	sourcePod       string
-	destinationPod  string
-	clientSet       kubernetes.Interface
-	smiAccessClient smiAccessClient.Interface
+	out               io.Writer
+	sourcePod         string
+	destinationPod    string
+	clientSet         kubernetes.Interface
+	smiAccessClient   smiAccessClient.Interface
+	configClient      osmConfigClient.Interface
```
Comment (shashankram) [inline cmd/cli/trafficpolicy_check.go:50]:
```suggestion
	meshConfigClient      osmConfigClient.Interface
```

File: cmd/cli/trafficpolicy_check.go [hunk 10]
```diff
@@ -33,17 +36,20 @@ osm policy check-pods bookbuyer-client bookstore-server
...
+	restConfig        *rest.Config
+	osmMeshConfigName string
```
Comment (shashankram) [inline cmd/cli/trafficpolicy_check.go:52]: This shouldn't be necessary if `--osm-namespace` is passed. Could you rely on that instead?

File: cmd/init-osm-controller/init-osm-controller.go [hunk 11]
```diff
@@ -55,11 +54,11 @@ func validateCLIParams() error {
 
 func createDefaultMeshConfig() *v1alpha1.MeshConfig {
 	return &v1alpha1.MeshConfig{
-		TypeMeta: v1.TypeMeta{
+		TypeMeta: metav1.TypeMeta{
```
Comment (shashankram) [inline cmd/init-osm-controller/init-osm-controller.go:57]: :+1:

File: cmd/osm-injector/osm-injector.go [hunk 12]
```diff
@@ -133,13 +134,14 @@ func main() {
 		metricsstore.DefaultMetricsStore.CertIssuedTime,
 	)
 
-	// Initialize Configurator to watch osm-config ConfigMap
-	cfg := configurator.NewConfigurator(kubeClient, stop, osmNamespace, osmConfigMapName)
-	configMap, err := cfg.GetConfigMap()
+	// This component will be watching the OSM MeshConfig and will make it
```
Comment (shashankram) [inline cmd/osm-injector/osm-injector.go:137]:
```suggestion
	// Initialize Configurator to retrieve mesh specific config
```

File: cmd/osm-injector/osm-injector.go [hunk 13]
```diff
@@ -133,13 +134,14 @@ func main() {
...
+	// to the rest of the components.
+	cfg := configurator.NewConfiguratorWithCRDClient(configClientset.NewForConfigOrDie(kubeConfig), stop, osmNamespace, osmMeshConfigName)
+	meshConfig, err := cfg.GetMeshConfigJSON()
 	if err != nil {
-		log.Error().Err(err).Msgf("Error parsing ConfigMap %s", osmConfigMapName)
+		log.Error().Err(err).Msgf("Error parsing MeshConfig %s", osmMeshConfigName)
 	}
-	log.Debug().Msgf("Initial ConfigMap %s: %s", osmConfigMapName, string(configMap))
+	log.Info().Msgf("Initial MeshConfig  %s: %v", osmMeshConfigName, meshConfig)
```
Comment (shashankram) [inline cmd/osm-injector/osm-injector.go:144]: We should not be dumping the entire MeshConfig resource. This could have user specific configuration values that might be sensitive. In the past we agreed to not do so.

File: manifests/mesh-config.yaml [hunk 14]
```diff
@@ -0,0 +1,21 @@
+apiVersion: config.openservicemesh.io/v1alpha1
```
Comment (shashankram) [inline manifests/mesh-config.yaml:1]: Could you move this example to `docs/example/manifests/` instead.

File: mockspec/generate.go [hunk 15]
```diff
@@ -72,6 +72,7 @@ func genMock(ruleStr string) {
 		"-w", ruleOptions[1],
 	}
 	cmd = exec.Command("go", cmdList...) // nolint gosec
+	log.Printf("command: go %v\n", cmdList)
```
Comment (shashankram) [inline mockspec/generate.go:75]: Remove, not related to this change.

File: pkg/apis/config/v1alpha1/mesh_config.go [hunk 16]
```diff
@@ -26,8 +26,8 @@ type SidecarSpec struct {
 	EnablePrivilegedInitContainer bool   `json:"enablePrivilegedInitContainer,omitempty" yaml:"enablePrivilegedInitContainer,omitempty"`
 	LogLevel                      string `json:"logLevel,omitempty" yaml:"logLevel,omitempty"`
 	EnvoyImage                    string `json:"envoyImage,omitempty" yaml:"envoyImage,omitempty"`
-	MaxDataPlaneConnections       int    `json:"maxMaxPlaneConnections,omitempty" yaml:"max_data_plane_connections,omitempty"`
-	ConfigResyncInterval          string `json:"configResyncInterval,omitempty" yaml:"config_resync_interval,omitempty"`
+	MaxDataPlaneConnections       int    `json:"maxDataPlaneConnections,omitempty" yaml:"maxDataPlaneConnections,omitempty"`
```
Comment (shashankram) [inline pkg/apis/config/v1alpha1/mesh_config.go:29]: Yaml tags are not required because the k8s API operates with JSON. `kubectl` converts YAML to JSON before making a request to the API server. Kindly remove the redundant tags since they seem unnecessary.

File: pkg/configurator/crd_client.go [hunk 17]
```diff
@@ -43,8 +103,7 @@ func newConfiguratorWithCRDClient(meshConfigClientSet versioned.Interface, stop
 	// start listener
 	go crdClient.runMeshConfigListener(stop)
 
-	// start informer
-	go informer.Run(stop)
+	crdClient.run(stop)
```
Comment (shashankram) [inline pkg/configurator/crd_client.go:106]: Is the change to drop the goroutine intentional? If yes, could you make this in a separate change with more context?

File: pkg/configurator/crd_client.go [hunk 18]
```diff
@@ -190,3 +250,83 @@ func meshConfigUpdatedMessageHandler(psubMsg *events.PubSubMessage) {
 			psubMsg.AnnouncementType)
 	}
 }
+
+func (c *CRDClient) getMeshConfigCacheKey() string {
+	return fmt.Sprintf("%s/%s", c.osmNamespace, c.meshConfigName)
+}
+
+// Returns the current ConfigMap
+func (c *CRDClient) getMeshConfig() *osmConfig {
```
Comment (shashankram) [inline pkg/configurator/crd_client.go:259]: I don't see why we need the `osmConfig` struct anymore, given that `MeshConfig` is already a well defined type. The intermediary type conversion seems unnecessary and was only necessary because the ConfigMap only supported string key-value pairs.

File: pkg/configurator/crd_client.go [hunk 19]
```diff
@@ -190,3 +250,83 @@ func meshConfigUpdatedMessageHandler(psubMsg *events.PubSubMessage) {
...
+	} else {
+		meshConfig = item.(*v1alpha1.MeshConfig)
+	}
+
+	return parseOSMMeshConfig(meshConfig)
+}
+
+// This struct must match the shape of the "osm-config" ConfigMap
+// which was created in the OSM namespace.
+type osmConfig struct {
```
Comment (shashankram) [inline pkg/configurator/crd_client.go:280]: Like noted above, this seems unnecessary. The MeshConfig object should be directly parseable.

File: pkg/configurator/crd_methods.go [hunk 20]
```diff
@@ -39,88 +39,88 @@ func (c *Client) GetConfigMap() ([]byte, error) {
...
-	return c.getConfigMap().PermissiveTrafficPolicyMode
+func (c *CRDClient) IsPermissiveTrafficPolicyMode() bool {
+	return c.getMeshConfig().PermissiveTrafficPolicyMode
 }
 
 // IsEgressEnabled determines whether egress is globally enabled in the mesh or not.
-func (c *Client) IsEgressEnabled() bool {
-	return c.getConfigMap().Egress
+func (c *CRDClient) IsEgressEnabled() bool {
+	return c.getMeshConfig().Egress
```
Comment (shashankram) [inline pkg/configurator/crd_methods.go:48]: As noted above, the conversion of `MeshConfig` -> `internal OsmConfig` is unnecessary. Can we just return the value from the MeshConfig struct instead?

**Comment by allenlsy** [comment]
@shashankram Hi thanks for helping with the review. I feel it's difficult further shrink the PR size, because pieces are intertwined. Without the complete change, the tests may not pass.

Replacing osm-config configmap affects:

* osm-controller
* osm-injector
* installation command and helm templates
* all related tests

File: pkg/configurator/crd_client.go (hunk 17 above)
Comment (allenlsy) [reply pkg/configurator/crd_client.go:106]: No, the goroutine invocation is moved into `crdClient.run()`

https://github.com/openservicemesh/osm/blob/c73d4b0a0195cab7ec5d1b1aac2d488ff5d52efa/pkg/configurator/crd_client.go#L149

File: pkg/configurator/crd_client.go (hunk 18 above)
Comment (allenlsy) [reply pkg/configurator/crd_client.go:259]: I agree. we can refactor this in a separate PR.

File: pkg/configurator/crd_methods.go (hunk 3 above)
Comment (allenlsy) [reply pkg/configurator/crd_methods.go:43]: This is similar to what @shashankram  mentioned, that we can remove the intermediate `osmconfig` struct. Will work on it in a separate PR.

File: pkg/configurator/client.go (hunk 6 above)
Comment (allenlsy) [reply pkg/configurator/client.go:19]: Will work on the refactor in a separate PR.

File: cmd/cli/trafficpolicy_check.go (hunk 10 above)
Comment (shashankram) [reply cmd/cli/trafficpolicy_check.go:52]: Let's keep the CLI behavior as before, and make such changes in a separate PR. So for this PR, I suggest to rely on the `--osm-namespace` flag because there is 1:1 mapping between a mesh and OSM's namespace.

Comment (allenlsy) [reply cmd/cli/trafficpolicy_check.go:52]: oh actually, the meshConfigName argument is optional. It falls back to `osm-mesh-config` if not provided. From the CLI perspective, the behavior is not changed.

I will update the trafficpolicy_check test cases to reflect it.

Comment (shashankram) [reply cmd/cli/trafficpolicy_check.go:52]: Can we hardcode the mesh-config name and update this CLI in a separate PR? Currently, the cli code hardcodes the ConfigMap name as `osm-config`.

File: cmd/cli/install.go [hunk 21]
```diff
@@ -91,6 +91,8 @@ var possibleEnvoyLogLevels = []string{"trace", "debug", "info", "warning", "warn
 // Its value is initialized at build time.
 var chartTGZSource string
 
+var log = logger.NewPretty("osm-install")
```
Comment (shashankram) [inline cmd/cli/install.go:94]: I don't know that we need a logger in the CLI. Why not just print to stdout/stderr?

File: cmd/cli/install.go [hunk 22]
```diff
@@ -293,6 +299,8 @@ func (i *installCmd) resolveValues() (map[string]interface{}, error) {
 }
 
 func (i *installCmd) validateOptions() error {
+	log.Info().Msgf("Start validateOptions: loadOSMChart")
```
Comment (shashankram) [inline cmd/cli/install.go:302]: This can be removed.

File: cmd/cli/trafficpolicy_check.go [hunk 23]
```diff
@@ -13,14 +13,17 @@ import (
...
-	"github.com/openservicemesh/osm/pkg/configurator"
+	osmConfigClient "github.com/openservicemesh/osm/pkg/gen/client/config/clientset/versioned"
 )
 
 const trafficPolicyCheckDescription = `
 This command will check whether a given source pod is allowed to communicate
 (send traffic) to a given destination pod by an SMI TrafficTarget policy or
 in lieu of the mesh operating in permissive traffic policy mode.
+
+MESH_CONFIG_NAME is optional. The default value is "osm-mesh-config".
```
Comment (shashankram) [inline cmd/cli/trafficpolicy_check.go:26]: When will this have a different name?

File: cmd/cli/trafficpolicy_check.go [hunk 24]
```diff
@@ -52,30 +58,42 @@ func newTrafficPolicyCheck(out io.Writer) *cobra.Command {
 	}
 
 	cmd := &cobra.Command{
-		Use:   "check-pods SOURCE_POD DESTINATION_POD",
+		Use:   "check-pods SOURCE_POD DESTINATION_POD [MESH_CONFIG_NAME]",
```
Comment (shashankram) [inline cmd/cli/trafficpolicy_check.go:61]: Let this not be a positional argument, instead define a flag `--mesh-config-name`

File: cmd/cli/trafficpolicy_check.go [hunk 25]
```diff
@@ -184,16 +202,13 @@ func (cmd *trafficPolicyCheckCmd) getMeshedPod(namespace, podName string) (*core
...
-	if err != nil {
-		return false, errors.Errorf("Error checking if permissive mode is enabled: %s", err)
-	}
 
-	configVal, err := configurator.GetBoolValueForKey(configMap, configurator.PermissiveTrafficPolicyModeKey)
+	meshConfig, err := cmd.meshConfigClient.ConfigV1alpha1().MeshConfigs(osmNamespace).Get(context.TODO(), cmd.osmMeshConfigName, metav1.GetOptions{})
+
 	if err != nil {
-		return false, errors.Errorf("Invalid value for key %q in %s/%s ConfigMap: %s", configurator.PermissiveTrafficPolicyModeKey, configMap.Namespace, configMap.Name, err)
+		return false, errors.Errorf("Fetching MeshConfig %s error: %s", cmd.osmMeshConfigName, err)
```
Comment (shashankram) [inline cmd/cli/trafficpolicy_check.go:209]:
```suggestion
		return false, errors.Errorf("Error fetching MeshConfig %s: %s", cmd.osmMeshConfigName, err)
```

File: pkg/configurator/crd_methods.go [hunk 26]
```diff
@@ -17,19 +17,19 @@ const (
 // The functions in this file implement the configurator.Configurator interface
```
Comment (shashankram) [inline pkg/configurator/crd_methods.go:17]: I don't see why the filename needs to be changed. Could you keep this as `methods.go`?

File: tests/framework/common.go [hunk 27]
```diff
@@ -1007,11 +995,50 @@ func (td *OsmTestData) AddNsToMesh(shouldInjectSidecar bool, ns ...string) error
 	return nil
 }
 
-// UpdateOSMConfig updates OSM configmap
-func (td *OsmTestData) UpdateOSMConfig(key, value string) error {
-	patch := []byte(fmt.Sprintf(`{"data": {%q: %q}}`, key, value))
-	_, err := td.Client.CoreV1().ConfigMaps(td.OsmNamespace).Patch(context.TODO(), "osm-config", types.StrategicMergePatchType, patch, metav1.PatchOptions{})
-	return err
+// CreateDefaultOSMConfig creates a MeshConfig object with default settings
+func (td *OsmTestData) CreateDefaultOSMConfig() *v1alpha1.MeshConfig {
```
Comment (shashankram) [inline tests/framework/common.go:999]: Why is this required? Who is calling it?

File: cmd/cli/trafficpolicy_check.go (hunk 10 above)
Comment (allenlsy) [reply cmd/cli/trafficpolicy_check.go:52]: Sure

File: cmd/cli/trafficpolicy_check.go (hunk 23 above)
Comment (allenlsy) [reply cmd/cli/trafficpolicy_check.go:26]: I messed up. I thought in the current implementation, configmap name can be customized.

Will remove.

File: cmd/cli/trafficpolicy_check.go (hunk 24 above)
Comment (allenlsy) [reply cmd/cli/trafficpolicy_check.go:61]: Similar as above. Will remove this.

File: pkg/configurator/crd_methods.go (hunk 26 above)
Comment (allenlsy) [reply pkg/configurator/crd_methods.go:17]: I initially thought this would be cleaner for code review.

But now seems keep methods.go makes more sense.

Will update.

File: tests/framework/common.go (hunk 27 above)
Comment (allenlsy) [reply tests/framework/common.go:999]: Forgot to remove.

This method was added when initially in the test cases, meshConfig object is created by `CreateDefaultOSMConfig`, rather than fetched from client.

File: cmd/cli/install.go [hunk 28]
```diff
@@ -201,16 +203,19 @@ func newInstallCmd(config *helm.Configuration, out io.Writer) *cobra.Command {
 }
 
 func (i *installCmd) run(config *helm.Configuration) error {
+	fmt.Print("Validating command options")
```
Comment (shashankram) [inline cmd/cli/install.go:206]: Previously I meant to say this is not really necessary apart from debugging. Could you remove this?

File: cmd/cli/install.go [hunk 29]
```diff
@@ -201,16 +203,19 @@ func newInstallCmd(config *helm.Configuration, out io.Writer) *cobra.Command {
...
 	}
 
+	log.Trace().Msgf("run(): start resolveValues()")
 	// values represents the overrides for the OSM chart's values.yaml file
 	values, err := i.resolveValues()
 	if err != nil {
 		return err
 	}
 
+	fmt.Print("Start installation using Helm charts")
```
Comment (shashankram) [inline cmd/cli/install.go:218]: This print doesn't seem to be useful either because Helm is the only supported way of installing OSM using `osm install`. I don't see the need to print additional stuff here at the moment. Could you pull this out into a separate PR if needed?

**Review by shashankram — APPROVED** [review]
Thanks for addressing the comments

File: cmd/cli/install.go (hunk 28 above)
Comment (allenlsy) [reply cmd/cli/install.go:206]: I felt the install command was too quiet :)

File: cmd/cli/install_test.go [hunk 30]
```diff
@@ -25,6 +26,10 @@ import (
...
+
```
Comment (snehachhabria) [inline cmd/cli/install_test.go:32]: nit : no need of a new line here. All osm packages can come in the same block

File: cmd/cli/install_test.go [hunk 31]
```diff
@@ -13,6 +13,7 @@ import (
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
 	tassert "github.com/stretchr/testify/assert"
+
```
Comment (snehachhabria) [inline cmd/cli/install_test.go:16]: nit : same here we do not need a new line between external packages

File: cmd/cli/trafficpolicy_check_test.go [hunk 32]
```diff
@@ -7,13 +7,15 @@ import (
...
-	fakeAccessClient "github.com/servicemeshinterface/smi-sdk-go/pkg/gen/client/access/clientset/versioned/fake"
+	fakeAccess "github.com/servicemeshinterface/smi-sdk-go/pkg/gen/client/access/clientset/versioned/fake"
 	tassert "github.com/stretchr/testify/assert"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes/fake"
 
-	"github.com/openservicemesh/osm/pkg/configurator"
+	fakeConfig "github.com/openservicemesh/osm/pkg/gen/client/config/clientset/versioned/fake"
+
```
Comment (snehachhabria) [inline cmd/cli/trafficpolicy_check_test.go:17]: nit : no spacing required between osm packages

File: pkg/configurator/methods.go [hunk 33]
```diff
@@ -161,18 +161,24 @@ func (c *Client) GetOutboundPortExclusionList() []string {
...
 	duration, err := time.ParseDuration(resyncDuration)
 	if err != nil {
 		log.Debug().Err(err).Msgf("Error parsing config resync interval: %s", duration)
 		return time.Duration(0)
 	}
 	return duration
 }
+
+// GetPermissiveTrafficPolicyMode returns the duration for resync interval.
+// If error or non-parsable value, returns 0 duration
```
Comment (snehachhabria) [inline pkg/configurator/methods.go:181]: the comments above this method are incorrect , please update it

File: cmd/cli/install.go (hunk 28 above)
Comment (shashankram) [reply cmd/cli/install.go]: Let's resolve that as a part of https://github.com/openservicemesh/osm/issues/912 instead.

**Review by snehachhabria — COMMENTED** [review]
What about the deletion of this chart : https://github.com/openservicemesh/osm/blob/main/charts/osm/templates/osm-configmap.yaml . Since it will no longer be needed after this change.

Also all documentation needs to be updated accordingly, this can come in a separate PR

File: cmd/cli/install_test.go (hunk 30 above)
Comment (shashankram) [reply cmd/cli/install_test.go:32]: I suggested following this pattern where autogenerated packages are grouped separately. It is useful to catch dependency issues that might creep in from API code referring to other local packages. Let's keep this as-is for this reason.

**Comment by allenlsy** [comment]
@snehachhabria Yes on removing the osm-config.yaml. I think it comes back during rebasing.

The documentation update will be done on this https://github.com/openservicemesh/osm/issues/3040

File: pkg/configurator/methods.go (hunk 33 above)
Comment (allenlsy) [reply pkg/configurator/methods.go:181]: I didn't notice there is a `IsPermissiveTrafficPolicyMode()`, which serves the same purpose.

This method will be removed.

File: cmd/cli/install.go [hunk 34]
```diff
@@ -205,6 +207,7 @@ func (i *installCmd) run(config *helm.Configuration) error {
 		return err
 	}
 
+	log.Trace().Msgf("run(): start resolveValues()")
```
Comment (draychev) [inline cmd/cli/install.go:210]: @allenlsy would you be so kind to use logger package in this repo - initialized w/ the right string it provides a bit more context around these log lines:
https://github.com/openservicemesh/osm/blob/707593c3a291ec9ac4a3e92836d8776cb3d84ce3/pkg/envoy/rds/types.go#L4-L9

**Review by draychev — APPROVED** [review]

File: cmd/cli/install.go (hunk 34 above)
Comment (allenlsy) [reply cmd/cli/install.go:210]: Shashank mentioned [here](https://github.com/openservicemesh/osm/pull/3198#discussion_r624264762) that we have issue #912 for similar purpose. Will work on it separately.
